midi-magic/src/guitar/strumming.ts

```diff
@@ -0,0 +1,213 @@
+import { Note, OutputChannel } from "webmidi";
+import { Signal } from "@preact/signals";
+import { Chord } from "../harmony/scales";
+import {
+  getChordNotes as getGuitarChordNotes,
+  getPowerChordNotes,
+} from "./chords";
+
+type VelocityOpts = {
+  velocity?: number;
+  velocityRand?: number;
+};
+
+const velocityDefaults: Required<VelocityOpts> = {
+  velocity: 0.9,
+  velocityRand: 0.3,
+};
+
+const calcVelocity = (rand: number, options: Required<VelocityOpts>) =>
+  options.velocity - options.velocityRand * rand;
+
+type StrumOpts = {
+  strumSpeed?: number;
+  strumSpeedRand?: number;
+};
+
+const strumDefaults: Required<StrumOpts> = {
+  strumSpeed: 20,
+  strumSpeedRand: 10,
+};
+
+const calcStrumDelay = (
+  rand: number,
+  options: Required<StrumOpts>,
+  index: number
+) => `+${index * (options.strumSpeed - options.strumSpeedRand * rand)}`;
+
+export type Strumming = {
+  handleDown(): void;
+  handleUp(): void;
+};
+
+
+export const FullStrumming = (
+  activeChord: Signal<Chord>,
+  notesOut: Signal<OutputChannel | null>,
+  options: { velocityDecay?: number } & StrumOpts & VelocityOpts = {}
+): Strumming => {
+  const opts = {
+    ...velocityDefaults,
+    ...strumDefaults,
+    velocityDecay: 0.05,
+    ...options,
+  };
+  const strum = (notes: number[]) => {
+    notesOut.value?.sendAllNotesOff();
+    const varietyNumber = Math.random();
+    const baseVelocity = calcVelocity(varietyNumber, opts);
+    notes.forEach((chordNote, index) => {
+      notesOut.value?.sendNoteOn(
+        new Note(chordNote, {
+          attack: baseVelocity - opts.velocityDecay * index,
+        }),
```

Comment:
Should we add a `duration: ...` here and below?

I noticed that the chord notes continue playing infinitely if not stopped by a chord change.

The duration could be quite long if seems necessary.

Reply:
My approach was to model a real guitar: after you strum a chord, it doesn't _really_ stop ringing until you release your left hand. We could either try to add some long duration here or actually keep track of ringing notes and send note offs explicitly. I'm leaning more on keeping track of currently ringing notes.